wagtail/core/models.py

```diff
@@ -36,9 +39,117 @@
+                "been disabled. To enable it, update '{0}' to use a backend "
+                "that actually stores and returns stored values, such as "
+                "Memcached or Redis. If multiple CACHES are configured, you"
+                "can use the WAGTAIL_SITE_CACHE setting to specify a cache "
+                "other than '{0}' to use instead".format(key))
+            return
+        if not settings.DEBUG and isinstance(cache, LocMemCache):
+            warnings.warn(
+                "WAGTAIL_SITE_CACHE is set to '{0}', which refers to a cache "
+                "that utilises Django's LocMemCache backend. Wagtail's site "
+                "caching feature is incompatible with LocMemCache when running "
+                "in a production environment, and so has been disabled. To "
+                "enable it, update your site to use a thread-safe backend, "
+                "such as Memcached or Redis. If multiple CACHES are configured, "
+                "you can use the WAGTAIL_SITE_CACHE setting to specify a "
+                "cache other than '{0}' to use instead".format(key))
+            return
+        return cache
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extracts 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple"""
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    def _get_keys_for_unique_hostname(self, hostname):
+        """Returns a tuple of additional keys to use when populating the cache
+        for a ``Site`` with a unique hostname. These keys allow requests to be
+        matched to the site even if the request's port value differs from the
+        site's saved ``port`` field value"""
+        valid_ports = getattr(settings, 'WAGTAIL_SITE_CACHE_VALID_PORTS',
+                              (80, 443, 8000, 8080, None))
+        return tuple("{0}:{1}".format(hostname, port) for port in valid_ports)
+
+    def clear_cache(self):
+        cache = self.get_site_cache()
+        if cache is not None:
+            cache.delete(SITE_CACHE_SITES_KEY)
+
+    def populate_cache(self):
+        """Populates the cache with a dictionary of ``Site`` objects, keyed
+        with hostname/port combinations, which get_for_request() will use
+        to match ``HttpRequest`` objects to the relevant ``Site``"""
+        cache = self.get_site_cache()
+        if cache is None:
+            return
+
+        sites = {}
+        unique_hostname_sites = {}
+        for site in self.all().select_related('root_page').iterator():
+
+            sites["{0}:{1}".format(site.hostname, site.port)] = site
```

Comment:
There are three places in this PR where the cache key is generated using some form of `"{0}:{1}".format(hostname, port)`. Does it make sense to encapsulate this in a method somewhere for easier lookup of the logic used to generate the key?

Reply:
I did have something similar to this before the latest refactor, but ditched it in favour of keeping things more readable. However, you are definitely right - ensuring these keys are handled consistently is important.